alf/trainers/evaluator.py

```diff
@@ -0,0 +1,253 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from absl import logging
+from absl import flags
+import queue
+import torch.multiprocessing as mp
+import traceback
+import os
+import sys
+import time
+import torch
+from typing import Optional
+
+import alf
+from alf.algorithms.rl_algorithm import RLAlgorithm
+from alf.environments.alf_environment import AlfEnvironment
+from alf.utils import common
+from alf.data_structures import StepType
+from alf.algorithms.data_transformer import create_data_transformer
+from alf.environments.utils import create_environment
+from alf.trainers import policy_trainer
+from collections import namedtuple
+
+EvalJob = namedtuple(
+    "EvalJob", ["type", "global_counter", "env_steps", "state_dict"],
+    defaults=[None] * 4)
+
+
+class AsyncEvaluator(object):
+    """Evaluate algorithm performance asynchronously in a different process.
+
+    Args:
+        conf_file: path of the config file
+        root_dir: directory for saving summary and checkpoints
+        num_evns: the number of environments for evaluating
+        seed: random seed
+    """
+
+    def __init__(self,
+                 conf_file: str,
+                 root_dir: str,
+                 num_envs: int,
+                 seed: Optional[int] = None):
+        ctx = mp.get_context('spawn')
+        self._job_queue = ctx.Queue()
+        self._done_queue = ctx.Queue()
+        self._worker = ctx.Process(
+            target=_worker,
+            args=(self._job_queue, self._done_queue, conf_file, num_envs,
+                  root_dir, seed))
+        self._worker.start()
+
+    def eval(self, algorithm: RLAlgorithm, current_env_steps: int):
+        """Do one round of evaluation.
+
+        This function will return once the evaluator worker makes a copy of the
+        state_dict of ``algorithm``. However, if the previous evaluation has not
+        been finished, it will wait until it is finished.
+
+        The evaluation result will be written to log file and tensorboard by the
+        evaluation worker.
+
+        Args:
+            algorithm: the training algorithm
+            current_env_steps: current total number of environment steps.
+        """
+        job = EvalJob(
+            type="eval",
+            env_steps=int(current_env_steps),
+            global_counter=int(alf.summary.get_global_counter()),
+            state_dict=algorithm.state_dict())
```

Comment:
This limited number of EvalJob fields seems to make `gen_summaries` with other metrics not useful. Might have a field that store a list of steps?

Reply:
Changed env_steps to step_metric_values.